README.md
[![Build Status](https://travis-ci.org/rbarton65/espnff.svg?branch=master)](https://travis-ci.org/rbarton65/espnff) [![version](https://img.shields.io/badge/version-1.3.1-blue.svg)](https://github.com/rbarton65/espnff/blob/master/CHANGELOG.md) [![PyPI version](https://badge.fury.io/py/espnff.svg)](https://badge.fury.io/py/espnff)

# ESPN Fantasy Football API

Using the Espn API in combination with the NFL V1 API, this program gathers data from your fantasy football league,
both private and public. You can then visualize data using Dash by Plotly.

## Getting Started

The application can run from a web platform without downloading Python, but you can also use it for development purposes.

### Installing
With pip:

```python3
pip3 install espnff
```

With Git:

```bash
git clone https://github.com/rbarton65/espnff

cd espnff

python3 setup.py install
```


## Basic Usage

This gives an overview of all the features of `espnff`

### Downloading a public league

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league
League 123456, 2016 Season
```

### Viewing teams in a public league

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.teams
[Team(Team 1), Team(Team 2), Team(Team 3), Team(Team 4),
Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)]
>>> team1 = league.teams[0]
>>> team1
Team(Team 1)
```

### Viewing data for specific team

```python3
>>> league.teams
[Team(Team 1), Team(Team 2), Team(Team 3), Team(Team 4),
Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)]
>>> team1 = league.teams[0]
>>> team1.team_id
1
>>> team1.team_name
Team 1
>>> team1.team_abbrev
T1
>>> team1.owner
Roger Goodell
>>> team1.division_id
0
>>> team1.division_name
Division 1
>>> team1.wins
5
>>> team1.losses
1
>>> team1.points_for
734.69
>>> team1.points_against
561.15
>>> team1.schedule
[Team(Team 2), Team(Team 3), Team(Team 4), Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8),
Team(Team 2), Team(Team 3), Team(Team 4), Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)
>>> team1.scores
[135.5, 126.38, 129.53, 126.65, 114.81, 101.82, 1.15, 0, 0, 0, 0, 0, 0, 0]
>>> team1.mov
[32.12, 24.92, 45.97, 34.17, 41.74, -5.39, 1.15, 0, 0, 0, 0, 0, 0, 0]
```

### Viewing league settings

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> settings = league.settings
>>> settings
'Settings(League Name)'
>>> settings.reg_season_count
14
>>> self.final_season_count
16
>>> settings.undroppable_list
true
>>> settings.veto_votes_required
4
>>> settings.team_count
8
>>> settings.playoff_team_count
4
>>> settings.id
123456
>>> settings.keeper_count
3
>>> settings.tie_rule
'Most Bench Points'
>>> settings.playoff_seed_tie_rule
'Head to Head Record'
>>> settings.roster
{'RB/WR/TE': 1, 'BE': 7, 'QB': 1, 'D/ST': 1, 'RB': 2, 'TE': 1, 'K': 1, 'WR': 2}
>>> settings.trade_deadline
2016-11-16T17:00:00.000Z
>>> settings.name
League Name
>>> settings.status
playoffs
>>> settings.year
2016
>>> settings.server_date
2016-12-08T21:06:53.087Z
```

### Viewing power rankings

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.power_rankings(week=5)
[('31.85', Team(Team 1)), ('25.60', Team(Team 3)), ('25.60', Team(Team 6)), ('22.45', Team(Team 2)),
('20.70', Team(Team 8)), ('18.20', Team(Team 7)), ('18.20', Team(Team 4)), ('18.10', Team(Team 5))]
```

### Viewing scoreboard

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.scoreboard() # grab current week
["Matchup(Team(Team 2), Team(Team 7))", "Matchup(Team(Team 1), Team(Team 11))",
"Matchup(Team(Team 6), Team(Team 9))", "Matchup(Team(Team 12), Team(Team 4))",
"Matchup(Team(Team 10), Team(Team 3))", "Matchup(Team(Team 8), Team(Team 5))"]
>>> scoreboard = league.scoreboard(week=12) # define week
>>> scoreboard
["Matchup(Team(Team 2), Team(Team 7))", "Matchup(Team(Team 1), Team(Team 11))",
"Matchup(Team(Team 6), Team(Team 9))", "Matchup(Team(Team 12), Team(Team 4))",
"Matchup(Team(Team 10), Team(Team 3))", "Matchup(Team(Team 8), Team(Team 5))"]
>>> matchup = scoreboard[1]
>>> matchup
"Matchup(Team(Team 1), Team(Team 11))"
>>> matchup.home_team
"Team(Team 1)"
>>> matchup.home_score
7.05
>>> matchup.away_team
"Team(Team 11)"
>>> matchup.away_score
45.85
```

## Running the tests

Automated tests for this package are included in the `tests` directory. After installation,
you can run these tests by changing the directory to the `espnff` directory and running the following:

```python3
python3 setup.py test
```

## Versioning

This library uses [SemVer](http://semver.org/) for versioning. For available versions, see the
[tags on this repository](https://github.com/rbarton65/espnff/tags)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.



## [Unreleased]
### Patched
- Quick fix for `trade_deadline` attribute in `Settings` class. Will search for better alternative.

### Other
- Reorganized structure of `espnff` package for future development

## [1.3.1] - 2017-07-24
### Added
- Added CI testing for Python 3.6

### Patched
- Fixed issue where map for `playoff_seed_tie_rule` was wrong.

## [1.3.0] - 2016-12-12
### Added
- Added `settings` attribute for league instance.
- Added `roster`, `tie_rule`, and `playoff_seed_tie_rule` to `settings`

## [1.2.3] - 2016-12-06
### Patched
- Fixed issue where bye weeks were breaking `scoreboard()` function.

## [1.2.2] - 2016-11-30
### Patched
- Fixed `power_rankings()` issue where execution would print matrices.

## [1.2.1] - 2016-11-29
### Patched
- Fixed `power_rankings()` issue where lists were out of range.

## [1.2.0] - 2016-11-25
### Added
- Added scoreboard function to League instance.
- Added tests for scoreboard function.

### Removed
- Removed deprecation warnings, fully converted to new attributes.
- Removed need for `numpy` package.


## [1.1.0] - 2016-10-18
### Added
- `.travis.yml` file for CI testing.
- `setup.cfg` file for `flake8` testing.
- Http requests mock for testing.
- `league.status_code` for testing purposes.
- Exceptions for status codes.
- Tests for future development.

### Changed
- Deprecation warnings for `Team` class attributes. Next Version
will remove old attributes.
- README.md shows build status from Travis CI.
- Better documentation in README.md.
- Added range for dependencies.
- `from espnff import *` now only imports `League` and `Team`.


## [1.0.1] - 2016-10-18
### Added
- Tests for utility functions.

### Changed
- Output for matrix functions now lists instead of numpy array.

### Fixed
- Fix issue where not all teams showed in power rankings.
- Fix issue where power rankings output was not sorted.

## [1.0.0] - 2016-10-04
### Added
- New attributes for Teams.
- Started using Semantic Versioning.

### Changed
- Replaced urls with ESPN's private API url.
- Replaced `Members` class with `Teams`.
- `get_week` function now `power_rankings`
- Teams now accessable as list attribute in `League`.

### Removed
- `get_member`
- `get_all_members`

## 0.2.2 - 2016-09-27
### Fixed
- Fix issue where margin of victory calculation in the power rankings
algorithm accounted for 50% of the total power points. Should have been 5%.

## 0.2.1 - 2016-09-15
### Fixed
- Fix issue where scores from games that weren't played would become
a string character instead of an integer.

## 0.2.0 - 2016-07-03
### Changed
- Rearranged output for `get_week`.

## 0.1.0 - 2016-05-13
### Added
- Power rankings function `get_week` to `League` class.
- README information.
- `setup.py` to turn project into module.

### Changed
- Renamed `power_rankings.py` to `espnff.py`.

## 0.0.0 - 2016-05-12
### Added
- League and Members classes to obtain information about ESPN league.


[Unreleased]: https://github.com/rbarton65/espnff/compare/v1.3.1...HEAD
[1.3.1]: https://github.com/rbarton65/espnff/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/rbarton65/espnff/compare/v1.2.3...v1.3.0
[1.2.3]: https://github.com/rbarton65/espnff/compare/v1.2.2...v1.2.3
[1.2.2]: https://github.com/rbarton65/espnff/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/rbarton65/espnff/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/rbarton65/espnff/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/rbarton65/espnff/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/rbarton65/espnff/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/rbarton65/espnff/releases/tag/v1.0.0

tests/test_utils.py
import unittest


from bin.utils import (
    square_matrix,
    add_matrix,
    two_step_dominance)


class UtilsTestCase(unittest.TestCase):
    '''Tests for utils'''

    def test_square_matrix(self):
        '''Is the logic for square_matrix successful?'''

        matrix1 = matrix1_answer = [[1, 0], [0, 1]]
        self.assertEqual(square_matrix(matrix1), matrix1_answer)

    def test_add_matrix(self):
        '''Is the logic for add_matrix successful?'''

        matrix1 = [[1, 0], [0, 1]]
        matrix1_answer = [[2, 0], [0, 2]]
        self.assertEqual(add_matrix(matrix1, matrix1), matrix1_answer)

    def test_two_step_dominance(self):
        '''Is the logic for two_step_dominance successful?'''

        matrix = [[1, 0], [0, 1]]
        matrix_answer = [2, 2]
        self.assertEqual(two_step_dominance(matrix), matrix_answer)


if __name__ == '__main__':
    unittest.main()


tests/test_league.py
import requests_mock
import unittest
import json


from bin.league import (League)
from bin.exception import (PrivateLeagueException,
                           InvalidLeagueException,
                           UnknownLeagueException, )


class LeagueTestCase(unittest.TestCase):
    '''Test League class'''

    @requests_mock.Mocker()
    def test_public_league(self, m):
        '''Is the status code 200 when calling a public league?'''
        data = json.loads(open('tests/test_league.json').read())
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016', status_code=200, json=data)
        league = League(1234, 2016)
        self.assertEqual(league.status, 200)

    @requests_mock.Mocker()
    def test_private_league(self, m):
        '''Does the correct exception raise when the status code is 401?'''
        data = {"error": [{"message": "No permission to view this league", "code": "functional"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016', status_code=401, json=data)
        with self.assertRaises(PrivateLeagueException):
            League(1234, 2016)

    @requests_mock.Mocker()
    def test_invalid_league(self, m):
        '''Does the correct exception raise when the status code is 404?'''
        data = {"error": [{"message": "Unable to retrieve league (2016:1234)", "code": "functional"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016', status_code=404, json=data)
        with self.assertRaises(InvalidLeagueException):
            League(1234, 2016)

    @requests_mock.Mocker()
    def test_unknown_error(self, m):
        '''Does the correct exception raise when the status code is not 200, 401, or 404?'''
        data = {"error": [{"message": "java.lang.NullPointerException", "code": "exception"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016', status_code=500, json=data)
        with self.assertRaises(UnknownLeagueException):
            League(1234, 2016)


if __name__ == '__main__':
    unittest.main()


tests/test_scoreboard.py
import requests_mock
import unittest
import json


from bin.league import (League)
from bin.exception import (PrivateLeagueException,
                           InvalidLeagueException,
                           UnknownLeagueException, )


class ScoreboardTestCase(unittest.TestCase):
    '''Test Scoreboard funtion'''

    def setUp(self):
        self.data = json.loads(open('tests/test_league.json').read())
        self.scoreboard_data = json.loads(open('tests/test_scoreboard.json').read())
        self.scoreboard_expected = ["Matchup(Team(Team 2), Team(Team 7))", "Matchup(Team(Team 1), Team(Team 11))",
                                    "Matchup(Team(Team 6), Team(Team 9))", "Matchup(Team(Team 12), Team(Team 4))",
                                    "Matchup(Team(Team 10), Team(Team 3))", "Matchup(Team(Team 8), Team(Team 5))"]

    @requests_mock.Mocker()
    def test_public_scoreboard(self, m):
        '''Is the status code 200 when calling a public league?'''
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016',
              status_code=200, json=self.scoreboard_data)
        league = League(1234, 2016)
        league.scoreboard()
        self.assertEqual(league.status, 200)

    @requests_mock.Mocker()
    def test_private_scoreboard(self, m):
        '''Does the correct exception raise when the status code is 401?'''
        scoreboard_error = {"error": [{"message": "No permission to view this league", "code": "functional"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016',
              status_code=401, json=scoreboard_error)
        with self.assertRaises(PrivateLeagueException):
            league = League(1234, 2016)
            league.scoreboard()

    @requests_mock.Mocker()
    def test_invalid_scoreboard(self, m):
        '''Does the correct exception raise when the status code is 404?'''
        scoreboard_error = {"error": [{"message": "Unable to retrieve league (2016:1234)", "code": "functional"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016',
              status_code=404, json=scoreboard_error)
        with self.assertRaises(InvalidLeagueException):
            league = League(1234, 2016)
            league.scoreboard()

    @requests_mock.Mocker()
    def test_unknown_scoreboard(self, m):
        '''Does the correct exception raise when the status code is 500?'''
        scoreboard_error = {"error": [{"message": "java.lang.NullPointerException", "code": "functional"}]}
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016',
              status_code=500, json=scoreboard_error)
        with self.assertRaises(UnknownLeagueException):
            league = League(1234, 2016)
            league.scoreboard()

    @requests_mock.Mocker()
    def test_scoreboard_week(self, m):
        '''Does the correct url generate from params when a week is entered?'''
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016&matchupPeriodId=12',
              status_code=200, json=self.scoreboard_data)
        league = League(1234, 2016)
        league.scoreboard(week=12)
        self.assertEqual(league.status, 200)

    @requests_mock.Mocker()
    def test_scoreboard(self, m):
        '''Does the correct output return?'''
        m.get('http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=1234&seasonId=2016',
              status_code=200, json=self.data)
        m.get('http://games.espn.com/ffl/api/v2/scoreboard?leagueId=1234&seasonId=2016',
              status_code=200, json=self.scoreboard_data)
        league = League(1234, 2016)
        scoreboard = league.scoreboard()
        scoreboard = [str(x) for x in scoreboard]
        self.assertEqual(scoreboard, self.scoreboard_expected)


if __name__ == '__main__':
    unittest.main()


espnff/team.py
import requests
from .exception import UnknownLeagueException
from .player import Player
class Team(object):
    '''Teams are part of the league'''
    def __init__(self, data, leagueId, seasonId):
        self.team_id = data['teamId']
        self.league_id = leagueId
        self.season_id = seasonId
        self.team_abbrev = data['teamAbbrev']
        self.team_name = "%s %s" % (data['teamLocation'], data['teamNickname'])
        self.division_id = data['division']['divisionId']
        self.division_name = data['division']['divisionName']
        self.wins = data['record']['overallWins']
        self.losses = data['record']['overallLosses']
        self.points_for = data['record']['pointsFor']
        self.points_against = data['record']['pointsAgainst']
        self.owner = "%s %s" % (data['owners'][0]['firstName'],
                                data['owners'][0]['lastName'])
        self.schedule = []
        self.scores = []
        self.mov = []
        self._fetch_schedule(data)
        self.roster = self.get_roster()


    def __repr__(self):
        return 'Team(%s)' % (self.team_name, )

    def _fetch_schedule(self, data):
        '''Fetch schedule and scores for team'''
        matchups = data['scheduleItems']

        for matchup in matchups:
            if not matchup['matchups'][0]['isBye']:
                if matchup['matchups'][0]['awayTeamId'] == self.team_id:
                    score = matchup['matchups'][0]['awayTeamScores'][0]
                    opponentId = matchup['matchups'][0]['homeTeamId']
                else:
                    score = matchup['matchups'][0]['homeTeamScores'][0]
                    opponentId = matchup['matchups'][0]['awayTeamId']
            else:
                score = matchup['matchups'][0]['homeTeamScores'][0]
                opponentId = matchup['matchups'][0]['homeTeamId']

            self.scores.append(score)
            self.schedule.append(opponentId)

    def get_roster(self):
        '''Get roster for a given week'''
        params = {
            "leagueId": self.league_id,
            'seasonId': self.season_id,
            'teamId': self.team_id
        }

        cookies = {
            'espn_s2': "AEAiHk9BLrd3k24bECp7EdRWKIUMi7blgeQCRyDeHpVMtz1AKSQQhXAaymmp2JMoDQ941uS6R6opM4lu%2FblNkgsFS3MSVqPgVWZEAKH0EsQTq15LP0eP2YKOM%2BtEAuDXwQmifpLHRpjSrNPUDqp%2BXQatuQ0eKnrePlG50Vr%2Ft01nPXoo0%2BgpmD1LVFmRnKCX3J%2FKb%2BdDQZifeI3El4%2FV7gPlpgWDIGQ023RzUod0tpUJkZjMQQjaTmRqG88fypDR9Q2paczm%2Fqj3kwmXiaRdKnSp",
            'SWID': "{F09B888C-9941-4A01-9B88-8C99417A0114}"
        }
        newRequest = requests.get('http://games.espn.com/ffl/api/v2/boxscore', params=params, cookies = cookies)
        status = newRequest.status_code
        roster = []
        if status != 200:
            raise UnknownLeagueException('Unknown Error Getting roster for team ' + self.team_id % status)
        else:
            requestData = newRequest.json()
            for team in requestData['boxscore']['teams']:
                if team['teamId'] == self.team_id:
                    for player in team['slots']:
                        if player.get('player', 0) != 0:
                            roster.append(Player(player))
        return roster


espnff/league.py
import requests

from .utils import (two_step_dominance,
                    power_points, )
from .team import Team
from .settings import Settings
from .matchup import Matchup
from .exception import (PrivateLeagueException,
                        InvalidLeagueException,
                        UnknownLeagueException, )


class League(object):
    '''Creates a League instance for Public ESPN league'''
    def __init__(self, league_id, year, espn_s2=None, swid=None):
        self.league_id = league_id
        self.year = year
        self.ENDPOINT = "http://games.espn.com/ffl/api/v2/"
        self.teams = []
        self.espn_s2 = espn_s2
        self.swid = swid
        self.athlete_pool = []
        self._fetch_league()

    def __repr__(self):
        return 'League(%s, %s)' % (self.league_id, self.year, )

    def _fetch_league(self):
        params = {
            'leagueId': self.league_id,
            'seasonId': self.year
        }

        cookies = None
        if self.espn_s2 and self.swid:
            cookies = {
                'espn_s2': self.espn_s2,
                'SWID': self.swid
            }

        r = requests.get('%sleagueSettings' % (self.ENDPOINT, ), params=params, cookies=cookies)
        self.status = r.status_code
        data = r.json()

        if self.status == 401:
            raise PrivateLeagueException(data['error'][0]['message'])

        elif self.status == 404:
            raise InvalidLeagueException(data['error'][0]['message'])

        elif self.status != 200:
            raise UnknownLeagueException('Unknown %s Error' % self.status)

        self._fetch_teams(data)
        self._fetch_settings(data)
        self.athlete_pool = self._fetch_athlete_pool()

    def _fetch_teams(self, data):
        '''Fetch teams in league'''
        teams = data['leaguesettings']['teams']

        for team in teams:
            self.teams.append(Team(teams[team], data['metadata']['leagueId'], data['metadata']['seasonId']))

        # replace opponentIds in schedule with team instances
        for team in self.teams:
            for week, matchup in enumerate(team.schedule):
                for opponent in self.teams:
                    if matchup == opponent.team_id:
                        team.schedule[week] = opponent

        # calculate margin of victory
        for team in self.teams:
            for week, opponent in enumerate(team.schedule):
                mov = team.scores[week] - opponent.scores[week]
                team.mov.append(mov)

        # sort by team ID
        self.teams = sorted(self.teams, key=lambda x: x.team_id, reverse=False)

    def _fetch_settings(self, data):
        self.settings = Settings(data)

    def power_rankings(self, week):
        '''Return power rankings for any week'''

        # calculate win for every week
        win_matrix = []
        teams_sorted = sorted(self.teams, key=lambda x: x.team_id,
                              reverse=False)

        for team in teams_sorted:
            wins = [0]*32
            for mov, opponent in zip(team.mov[:week], team.schedule[:week]):
                opp = int(opponent.team_id)-1
                if mov > 0:
                    wins[opp] += 1
            win_matrix.append(wins)
        dominance_matrix = two_step_dominance(win_matrix)
        power_rank = power_points(dominance_matrix, teams_sorted, week)
        return power_rank

    def scoreboard(self, week=None):
        '''Returns list of matchups for a given week'''
        params = {
            'leagueId': self.league_id,
            'seasonId': self.year
        }
        if week is not None:
            params['matchupPeriodId'] = week

        r = requests.get('%sscoreboard' % (self.ENDPOINT, ), params=params)
        self.status = r.status_code
        data = r.json()

        if self.status == 401:
            raise PrivateLeagueException(data['error'][0]['message'])

        elif self.status == 404:
            raise InvalidLeagueException(data['error'][0]['message'])

        elif self.status != 200:
            raise UnknownLeagueException('Unknown %s Error' % self.status)

        matchups = data['scoreboard']['matchups']
        result = [Matchup(matchup) for matchup in matchups]

        for team in self.teams:
            for matchup in result:
                if matchup.home_team == team.team_id:
                    matchup.home_team = team
                if matchup.away_team == team.team_id:
                    matchup.away_team = team

        return result

    def _fetch_athlete_pool(self):
        pool_url = ""
        pool_request = requests.get("http://api.espn.com/v1/sports/football/nfl/athletes")
        pool_status = pool_request.status_code
        return pool_request.json()



espnff/player.py
class Player(object):
    def __init__(self, player):
        self.pvoRank = self.__setStat(player,'pvoRank')
        self.isDroppable = player['player']['droppable']
        self.percentStarted = player['player']['percentStarted']
        self.jerseryNum = self.__setStat(player['player'],'jersey')
        self.playerRatingSeason = self.__setStat(player['player'],'playerRatingSeason')
        self.isIREligible = player['player']['isIREligible']
        self.draftRank = player['player']['draftRank']
        self.firstName = player['player']['firstName']
        self.lastName = player['player']['lastName']
        self.healthStatus = player['player']['healthStatus']
        self.percentOwned = player['player']['percentOwned']
        self.proTeam_id = player['player']['proTeamId']
        self.isActive = player['player']['isActive']
        self.player_id = player['player']['playerId']
        self.percentChange = player['player']['percentChange']
        self.defaultPositionId = player['player']['defaultPositionId']
        self.__position = self.__setPosition()
        self.totalPoints = self.__setStat(player['player'],'totalPoints')
        self.isOnWatchList = player['watchList']
        self.isTradeLocked = player['isTradeLocked']
        self.projectedPoints = self.__setStat(player['currentPeriodProjectedStats'],'appliedStatTotal')

    def __repr__(self):
        return self.firstName + " " + self.lastName

    def __setStat(self, dict, key):
        if dict.get(key, 0) == 0:
            return None
        else:
            return dict[key]

    def __setPosition(self):
        positionAssignment = {
            1: "QB",
            2: "RB",
            3: "WR",
            4: "TE",
            5: "K",
            11: "LB",
            10: "EDR",
            12: "CB",
            13: "Safety",

        }
        return positionAssignment.get(self.defaultPositionId, self.defaultPositionId)

    def getPosition(self):
        return self.__position


espnff/settings.py
class Settings(object):
    '''Creates Settings object'''
    def __init__(self, data):
        self.reg_season_count = data['leaguesettings']['finalRegularSeasonMatchupPeriodId']
        self.undroppable_list = data['leaguesettings']['usingUndroppableList']
        self.veto_votes_required = data['leaguesettings']['vetoVotesRequired']
        self.team_count = data['leaguesettings']['size']
        self.final_season_count = data['leaguesettings']['finalMatchupPeriodId']
        self.playoff_team_count = data['leaguesettings']['playoffTeamCount']
        self.id = data['leaguesettings']['id']
        self.keeper_count = data['leaguesettings']['futureKeeperCount']
        try:
            self.trade_deadline = data['leaguesettings']['tradeDeadline']
        except:
            self.trade_deadline = 'Unknown'
        self.name = data['leaguesettings']['name']
        self.status = data['metadata']['status']
        self.year = data['metadata']['seasonId']
        self.server_date = data['metadata']['serverDate']
        self._fetch_roster_settings(data)
        self._fetch_tie_rules(data)

    def __repr__(self):
        return 'Settings(%s)' % (self.name)

    def _fetch_roster_settings(self, data):
        '''Grabs roster settings'''
        roster_map = {
                      0: 'QB',
                      1: 'TQB',
                      2: 'RB',
                      3: 'RB/WR',
                      4: 'WR',
                      5: 'WR/TE',
                      6: 'TE',
                      7: 'OP',
                      8: 'DT',
                      9: 'DE',
                      10: 'LB',
                      11: 'DL',
                      12: 'CB',
                      13: 'S',
                      14: 'DB',
                      15: 'DP',
                      16: 'D/ST',
                      17: 'K',
                      18: 'P',
                      19: 'HC',
                      20: 'BE',
                      21: 'IR',
                      22: '',
                      23: 'RB/WR/TE',
                      24: ''
                      }

        roster = data['leaguesettings']['slotCategoryItems']
        self.roster = {roster_map[i['slotCategoryId']]: i['num'] for i in roster
                       if i['num'] != 0}

    def _fetch_tie_rules(self, data):
        '''Grabs matchup and playoff seeding tie info'''
        tie_map = {
                   0: 'None',
                   1: 'Home Team Wins',
                   2: 'Most Bench Points',
                   3: 'Most QB Points',
                   4: 'Most RB Points'
                  }

        tie_id = data['leaguesettings']['tieRule']

        try:
            self.tie_rule = tie_map[tie_id]
        except:
            self.tie_rule = 'Unknown'

        playoff_tie_map = {
                           -1: 'Head to Head Record',
                           0: 'Total Points For',
                           1: 'Intra Division Record',
                           2: 'Total Points Against'
                          }

        playoff_id = data['leaguesettings']['playoffSeedingTieRuleRawStatId']

        try:
            self.playoff_seed_tie_rule = playoff_tie_map[playoff_id]
        except:
            self.playoff_seed_tie_rule = 'Unknown'
